trajopt_sco/include/trajopt_sco/osqp_interface.hpp

```diff
@@ -10,6 +10,25 @@ TRAJOPT_IGNORE_WARNINGS_POP
 
 namespace sco
 {
+struct OSQPWorkspaceInterface
+{
+  OSQPWorkspaceInterface() { data_ = nullptr; }
+  ~OSQPWorkspaceInterface()
+  {
+    if (data_ != nullptr)
+      osqp_cleanup(data_);
+  }
+  OSQPWorkspaceInterface(const OSQPWorkspaceInterface& model) { data_ = model.data_; }
```

Comment:
I think in this case it is valid to delete the copy and assignment operator and only allow a move operator. Then the operator would have to create a new object.

Reply:
Which at this point it is equivalent to making a unique pointer of OSQPWorkspace but it calls a custom delete function. When you create the unique point you can pass in a custom deletor function to be called.

```
std::unique_ptr<OSQPWorkspace, [] (OSQPWorkspace* ptr) { osqp_cleanup(ptr); }>> w;
```

Reply:
I tested the code below in [godbolt](https://godbolt.org/) and each copy does point to the same object.
```
struct OSQPWorkspaceInterface
{
  OSQPWorkspaceInterface() { data_ = nullptr; }
  ~OSQPWorkspaceInterface()
  {
    std::cout << (void *)data_ << std::endl;
    delete data_;
  }
  OSQPWorkspaceInterface(const OSQPWorkspaceInterface& model) { data_ = model.data_; }
  OSQPWorkspaceInterface& operator=(const OSQPWorkspaceInterface& model)
  {
    data_ = model.data_;
    return *this;
  }
  OSQPWorkspaceInterface(OSQPWorkspaceInterface&&) = default;
  OSQPWorkspaceInterface& operator=(OSQPWorkspaceInterface&&) = default;
  double* data_;
};

int main(int , char** )
{
    
    OSQPWorkspaceInterface test;
    test.data_ = new double;

    OSQPWorkspaceInterface test2 = test;
}
```

Reply:
Yeah. I guess I would have expected it to. We are storing it as a unique_ptr, but OSQP presumably also has copies of the raw ptr around. I guess this would be more that you are assigning a new interface to the underlying workspace not so much assigning an interface to an existing one.

Regardless, I adding a custom destructor to the pointer is really what I was wanting to do, so I don't think this interface struct is needed.

Reply:
Ok, then you can do the same thing with shared pointers to have a custom deletor.